File: extensions/nexus-forum/src/Service/AgentPreflightCatalog.php
<?php

namespace Nexus\Forum\Service;

class AgentPreflightCatalog
{
    private const AGENT_PUBLIC_FOOTER = 'Posted by Nexus Agent after explicit user confirmation.';
    private const AGENT_HELP_FOOTER = 'Drafted by Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.';
    private const AGENT_MATCH_OFFER_FOOTER = 'Offered through Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.';
    private const AGENT_DISPATCH_ACCEPT_FOOTER = 'Accepted through Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.';

    private const PERMISSION_COLUMNS = [
        'allowAgentPosting' => 'allow_agent_posting',
        'allowAgentReplying' => 'allow_agent_replying',
        'allowAgentMatching' => 'allow_agent_matching',
        'allowLocationMatching' => 'allow_location_matching',
    ];

    public static function actions(): array
    {
        return self::catalog();
    }

    public static function actionNames(): array
    {
        return array_keys(self::catalog());
    }

    public static function hasAction(string $action): bool
    {
        return array_key_exists($action, self::catalog());
    }

    public static function definition(string $action): array
    {
        return self::catalog()[$action];
    }

    public static function permissionColumn(string $permission): string
    {
        return self::PERMISSION_COLUMNS[$permission];
    }

    public static function publicFooter(): string
    {
        return self::AGENT_PUBLIC_FOOTER;
    }

    public static function helpFooter(): string
    {
        return self::AGENT_HELP_FOOTER;
    }

    public static function matchOfferFooter(): string
    {
        return self::AGENT_MATCH_OFFER_FOOTER;
    }

    public static function dispatchAcceptFooter(): string
    {
        return self::AGENT_DISPATCH_ACCEPT_FOOTER;
    }

    private static function catalog(): array
    {
        return [
        'need_draft' => [
            'endpoint' => 'POST /api/nexus/need-drafts',
            'requiresConfirmation' => false,
            'permissions' => [],
            'purpose' => 'Classify a raw natural-language user need, return a draft and read-only discoveryPlan, and avoid publishing or writing database state.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['rawUserNeed', 'intent', 'locationHint'],
            'writeBodySchemaRef' => '#/components/schemas/NeedDraftInput',
            'sideEffects' => self::sideEffects([
                'writesDatabase' => false,
                'createsPublicContent' => false,
                'createsActionLog' => false,
                'visibility' => 'private_to_calling_agent_response',
            ]),
            'examplePreflightBody' => [
                'data' => [
                    'type' => 'nexus-agent-preflights',
                    'attributes' => [
                        'action' => 'need_draft',
                    ],
                ],
            ],
        ],
        'forum_discussion.create' => [
            'endpoint' => 'POST /api/nexus/forum/discussions',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentPosting'],
            'purpose' => 'Create a general forum discussion through the Nexus gateway after the user confirms the exact public title, body, and tags.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['title', 'content', 'tagIds', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/ForumDiscussionInput',
            'sideEffects' => self::sideEffects([
                'createsPublicContent' => true,
                'createsPublicDiscussion' => true,
                'visibility' => 'public_forum',
                'serverAppendedFooter' => self::AGENT_PUBLIC_FOOTER,
            ]),
            'examplePreflightBody' => self::example('forum_discussion.create'),
        ],
        'forum_post.reply' => [
            'endpoint' => 'POST /api/nexus/forum/discussions/{id}/posts',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentReplying'],
            'purpose' => 'Reply to an existing forum discussion through the Nexus gateway after the user confirms the exact public reply.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => ['discussion id in endpoint path'],
            'proposedFields' => ['content', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/ForumPostInput',
            'sideEffects' => self::sideEffects([
                'createsPublicContent' => true,
                'createsPublicReply' => true,
                'visibility' => 'public_forum',
                'serverAppendedFooter' => self::AGENT_PUBLIC_FOOTER,
            ]),
            'examplePreflightBody' => self::example('forum_post.reply'),
        ],
        'forum_post.edit' => [
            'endpoint' => 'PATCH /api/nexus/forum/posts/{id}',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentReplying'],
            'purpose' => 'Edit the current user\'s own forum post through the Nexus gateway after exact user confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => ['post id in endpoint path'],
            'proposedFields' => ['content', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/ForumPostInput',
            'sideEffects' => self::sideEffects([
                'updatesPublicContent' => true,
                'visibility' => 'public_forum',
                'serverAppendedFooter' => 'Edited by Nexus Agent after explicit user confirmation.',
            ]),
            'examplePreflightBody' => self::example('forum_post.edit'),
        ],
        'forum_post.delete' => [
            'endpoint' => 'DELETE /api/nexus/forum/posts/{id}',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentReplying'],
            'purpose' => 'Hide the current user\'s own forum post through the Nexus gateway after exact user confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => ['post id in endpoint path'],
            'proposedFields' => ['reason', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/ForumPostDeleteInput',
            'sideEffects' => self::sideEffects([
                'updatesPublicContent' => true,
                'softDeletesOwnPost' => true,
                'visibility' => 'public_forum_post_hidden',
            ]),
            'examplePreflightBody' => self::example('forum_post.delete'),
        ],
        'agent_profile.update' => [
            'endpoint' => 'PATCH /api/nexus/me/agent-profile',
            'requiresConfirmation' => true,
            'permissions' => [],
            'purpose' => 'Update the current user\'s private Agent Profile, soul.md metadata, preferences, or authorization switches after explicit confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['agentName', 'soulMd', 'interestTags', 'skillTags', 'helpTags', 'matchPreferences', 'permissions', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/AgentProfileInput',
            'sideEffects' => self::sideEffects([
                'visibility' => 'current_user_private_profile',
                'containsPrivateProfileData' => true,
            ]),
            'examplePreflightBody' => self::example('agent_profile.update'),
        ],
        'capabilities.update' => [
            'endpoint' => 'PATCH /api/nexus/me/capabilities',
            'requiresConfirmation' => true,
            'permissions' => [],
            'purpose' => 'Update the current user\'s public helper capability labels after explicit confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['capabilities', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/CapabilityCollectionInput',
            'sideEffects' => self::sideEffects([
                'updatesPublicProfile' => true,
                'visibility' => 'public_capability_labels',
            ]),
            'examplePreflightBody' => self::example('capabilities.update'),
        ],
        'help_request.create' => [
            'endpoint' => 'POST /api/nexus/help-requests',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Create a public real-world help request only after read-before-write discovery and exact requester confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['title', 'summary', 'content', 'neededLabels', 'categoryLabel', 'urgency', 'locationHint', 'meetingSafetyState', 'agentContext', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/HelpRequestInput',
            'sideEffects' => self::sideEffects([
                'createsHelpRequest' => true,
                'createsPublicContent' => true,
                'createsPublicDiscussion' => true,
                'visibility' => 'public_help_board_and_forum',
                'serverAppendedFooter' => self::AGENT_HELP_FOOTER,
            ]),
            'examplePreflightBody' => self::example('help_request.create'),
        ],
        'help_request.update' => [
            'endpoint' => 'PATCH /api/nexus/help-requests/{id}',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Update the requester-owned help request status, safe meeting state, public-safe location hint, summary, or labels.',
            'targetType' => 'help_request',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'helpRequestId'],
            'proposedFields' => ['status', 'meetingSafetyState', 'locationHint', 'summary', 'neededLabels', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/HelpRequestUpdateInput',
            'sideEffects' => self::sideEffects([
                'updatesHelpRequest' => true,
                'visibility' => 'public_help_metadata',
            ]),
            'examplePreflightBody' => self::targetExample('help_request.update', 'help_request', ['status' => 'closed']),
        ],
        'dispatch.create' => [
            'endpoint' => 'POST /api/nexus/help-requests/{id}/dispatches',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Invite one candidate helper to a requester-owned open help request after requester confirmation.',
            'targetType' => 'help_request',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'helpRequestId'],
            'proposedFields' => ['helperUserId', 'message', 'rationale', 'meetingHint', 'meetingSafetyState', 'expiresAt', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/HelpDispatchInput',
            'sideEffects' => self::sideEffects([
                'createsDispatch' => true,
                'createsNotification' => true,
                'visibility' => 'requester_helper_scoped_dispatch',
            ]),
            'examplePreflightBody' => self::targetExample('dispatch.create', 'help_request', ['helperUserId' => 2]),
        ],
        'dispatch.update' => [
            'endpoint' => 'PATCH /api/nexus/dispatches/{id}',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Accept, decline, or cancel a scoped dispatch invitation after the acting participant confirms the exact response.',
            'targetType' => 'help_dispatch',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'dispatchId'],
            'proposedFields' => ['status', 'responseMessage', 'message', 'meetingHint', 'meetingSafetyState', 'userConfirmed'],
            'allowedProposedStatus' => ['accepted', 'declined', 'cancelled'],
            'writeBodySchemaRef' => '#/components/schemas/HelpDispatchUpdateInput',
            'sideEffects' => self::sideEffects([
                'updatesDispatch' => true,
                'mayCreateMatch' => true,
                'mayCreatePublicReply' => true,
                'visibility' => 'requester_helper_scoped_dispatch',
                'serverAppendedFooterWhenAccepted' => self::AGENT_DISPATCH_ACCEPT_FOOTER,
            ]),
            'examplePreflightBody' => self::targetExample('dispatch.update', 'help_dispatch', ['status' => 'accepted']),
        ],
        'match.offer' => [
            'endpoint' => 'POST /api/nexus/help-requests/{id}/matches',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Offer help on another user\'s open help request after the helper confirms the exact public offer.',
            'targetType' => 'help_request',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'helpRequestId'],
            'proposedFields' => ['message', 'meetingHint', 'meetingSafetyState', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/MatchInput',
            'sideEffects' => self::sideEffects([
                'createsMatchOffer' => true,
                'createsPublicReply' => true,
                'visibility' => 'public_help_thread_and_scoped_match',
                'serverAppendedFooter' => self::AGENT_MATCH_OFFER_FOOTER,
            ]),
            'examplePreflightBody' => self::targetExample('match.offer', 'help_request', ['message' => 'I can help at a public campus spot.']),
        ],
        'match.update' => [
            'endpoint' => 'PATCH /api/nexus/matches/{id}',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Accept, decline, cancel, or complete a match after the acting participant confirms the exact state change.',
            'targetType' => 'help_match',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'matchId'],
            'proposedFields' => ['status', 'meetingHint', 'meetingSafetyState', 'userConfirmed'],
            'allowedProposedStatus' => ['offered', 'accepted', 'declined', 'cancelled', 'completed'],
            'writeBodySchemaRef' => '#/components/schemas/MatchUpdateInput',
            'sideEffects' => self::sideEffects([
                'updatesMatch' => true,
                'mayUpdateHelpRequestStatus' => true,
                'visibility' => 'requester_helper_scoped_match',
            ]),
            'examplePreflightBody' => self::targetExample('match.update', 'help_match', ['status' => 'accepted']),
        ],
        'match_message.create' => [
            'endpoint' => 'POST /api/nexus/matches/{id}/messages',
            'requiresConfirmation' => true,
            'permissions' => ['allowAgentMatching'],
            'purpose' => 'Send a private requester/helper coordination message only after the match is accepted and the user confirms the exact text.',
            'targetType' => 'help_match',
            'targetRequiredForPreflight' => true,
            'targetIdAliases' => ['target.id', 'targetId', 'matchId'],
            'proposedFields' => ['content', 'agentContext', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/MatchMessageInput',
            'sideEffects' => self::sideEffects([
                'createsPrivateMessage' => true,
                'visibility' => 'requester_helper_private_match_message',
            ]),
            'examplePreflightBody' => self::targetExample('match_message.create', 'help_match', ['content' => 'Let us meet at the library front desk.']),
        ],
        'device_signal.create' => [
            'endpoint' => 'POST /api/nexus/device-signals',
            'requiresConfirmation' => true,
            'permissions' => [],
            'purpose' => 'Upload a coarse, short-lived current-user device/location placeholder for future Linkgo/mobile integration after confirmation.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['purpose', 'coarseGeohash', 'accuracyM', 'bluetoothSeen', 'shakeDetected', 'gyroAvailable', 'payload', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/DeviceSignalInput',
            'sideEffects' => self::sideEffects([
                'storesShortLivedDeviceSignal' => true,
                'visibility' => 'current_user_private_device_signal',
            ]),
            'examplePreflightBody' => self::example('device_signal.create'),
        ],
        'llm_settings.update' => [
            'endpoint' => 'PATCH /api/nexus/llm-settings',
            'requiresConfirmation' => true,
            'permissions' => [],
            'purpose' => 'Update the current user\'s optional forum-side LLM provider metadata after confirmation; local agents can still call Nexus APIs without this.',
            'targetType' => null,
            'targetRequiredForPreflight' => false,
            'targetIdAliases' => [],
            'proposedFields' => ['provider', 'baseUrl', 'apiKey', 'chatModel', 'responsesModel', 'supportsChatCompletions', 'supportsResponses', 'userConfirmed'],
            'writeBodySchemaRef' => '#/components/schemas/LlmSettingsInput',
            'sideEffects' => self::sideEffects([
                'storesPrivateLlmSettings' => true,
                'storesWriteOnlyApiKey' => true,
                'visibility' => 'current_user_private_llm_settings',
            ]),
            'examplePreflightBody' => self::example('llm_settings.update'),
        ],
        ];
    }

    private static function sideEffects(array $overrides = []): array
    {
        return array_merge([
            'writesDatabase' => true,
            'createsPublicContent' => false,
            'createsPublicDiscussion' => false,
            'createsPublicReply' => false,
            'createsNotification' => false,
            'createsActionLog' => true,
            'visibility' => 'current_user_scoped',
        ], $overrides);
    }

    private static function example(string $action): array
    {
        return [
            'data' => [
                'type' => 'nexus-agent-preflights',
                'attributes' => [
                    'action' => $action,
                    'userConfirmed' => false,
                ],
            ],
        ];
    }

    private static function targetExample(string $action, string $targetType, array $proposed): array
    {
        $attributes = [
            'action' => $action,
            'userConfirmed' => false,
            'target' => [
                'type' => $targetType,
                'id' => 1,
            ],
        ];

        if ($proposed !== []) {
            $attributes['proposed'] = $proposed;
        }

        return [
            'data' => [
                'type' => 'nexus-agent-preflights',
                'attributes' => $attributes,
            ],
        ];
    }
}

File: server/src/main/java/nexus/campus/agent/service/AgentPreflightCatalog.java
package nexus.campus.agent.service;

import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class AgentPreflightCatalog {

    public static final String PUBLIC_FOOTER = "Posted by Nexus Agent after explicit user confirmation.";
    public static final String HELP_FOOTER = "Drafted by Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.";
    public static final String MATCH_OFFER_FOOTER = "Offered through Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.";
    public static final String DISPATCH_ACCEPT_FOOTER = "Accepted through Nexus Agent after explicit user confirmation. Offline coordination should prefer public, safe, easy-to-leave places.";

    private static final Map<String, String> PERMISSION_COLUMNS = Map.of(
        "allowAgentPosting", "allow_agent_posting",
        "allowAgentReplying", "allow_agent_replying",
        "allowAgentMatching", "allow_agent_matching",
        "allowLocationMatching", "allow_location_matching"
    );

    private final Map<String, Map<String, Object>> catalog;

    public AgentPreflightCatalog() {
        catalog = buildCatalog();
    }

    public Set<String> actionNames() { return catalog.keySet(); }
    public boolean hasAction(String action) { return catalog.containsKey(action); }
    public Map<String, Object> definition(String action) { return catalog.get(action); }
    public static String permissionColumn(String perm) { return PERMISSION_COLUMNS.get(perm); }

    private Map<String, Map<String, Object>> buildCatalog() {
        var c = new LinkedHashMap<String, Map<String, Object>>();

        c.put("need_draft", Map.of(
            "endpoint", "POST /api/nexus/need-drafts",
            "requiresConfirmation", false,
            "permissions", List.of(),
            "purpose", "Classify a raw natural-language user need, return a draft and read-only discoveryPlan, and avoid publishing or writing database state.",
            "proposedFields", List.of("rawUserNeed", "intent", "locationHint"),
            "sideEffects", Map.of(
                "writesDatabase", false,
                "createsPublicContent", false,
                "createsActionLog", false,
                "visibility", "private_to_calling_agent_response"
            )
        ));

        c.put("forum_discussion.create", Map.of(
            "endpoint", "POST /api/nexus/forum/discussions",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentPosting"),
            "purpose", "Create a general forum discussion through the Nexus gateway after the user confirms the exact public title, body, and tags.",
            "proposedFields", List.of("title", "content", "tagIds", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", true,
                "createsPublicDiscussion", true,
                "visibility", "public_forum",
                "serverAppendedFooter", PUBLIC_FOOTER
            )
        ));

        c.put("forum_post.reply", Map.of(
            "endpoint", "POST /api/nexus/forum/discussions/{id}/posts",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentReplying"),
            "purpose", "Reply to an existing forum discussion through the Nexus gateway after the user confirms the exact public reply.",
            "proposedFields", List.of("content", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", true,
                "createsPublicReply", true,
                "visibility", "public_forum",
                "serverAppendedFooter", PUBLIC_FOOTER
            )
        ));

        c.put("agent_profile.update", Map.of(
            "endpoint", "PATCH /api/nexus/me/agent-profile",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Update the current user's private Agent Profile, soul.md metadata, preferences, or authorization switches after explicit confirmation.",
            "proposedFields", List.of("agentName", "soulMd", "interestTags", "skillTags", "helpTags",
                    "matchPreferences", "permissions", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "private_to_user"
            )
        ));

        c.put("capabilities.update", Map.of(
            "endpoint", "PATCH /api/nexus/me/capabilities",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Update the current user's public capability labels after explicit confirmation.",
            "proposedFields", List.of("capabilities", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", true,
                "visibility", "public_capability_catalog"
            )
        ));

        c.put("memory.create", Map.of(
            "endpoint", "POST /api/nexus/me/memories",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Store a durable private memory after the user confirms the exact content, sensitivity, retention, and sharing policy.",
            "proposedFields", List.of("kind", "title", "content", "tags", "importance",
                    "pinned", "sourceType", "sourceRef", "sensitivity", "sharePolicy",
                    "validFrom", "expiresAt", "userConfirmed"),
            "sideEffects", Map.of(
                    "createsPublicContent", false,
                    "visibility", "private_to_user",
                    "credentialsAllowed", false)
        ));

        c.put("memory.update", Map.of(
            "endpoint", "PATCH /api/nexus/me/memories/{id}",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Correct, archive, reprioritize, or change the retention policy of a private memory after explicit confirmation.",
            "proposedFields", List.of("kind", "title", "content", "tags", "status",
                    "importance", "pinned", "sensitivity", "sharePolicy", "expiresAt",
                    "clearExpiresAt", "userConfirmed"),
            "sideEffects", Map.of(
                    "createsPublicContent", false,
                    "visibility", "private_to_user")
        ));

        c.put("memory.delete", Map.of(
            "endpoint", "DELETE /api/nexus/me/memories/{id}",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Permanently delete a private memory and its match-share snapshots after explicit confirmation.",
            "proposedFields", List.of("userConfirmed"),
            "sideEffects", Map.of(
                    "destructive", true,
                    "visibility", "private_to_user")
        ));

        c.put("memory_share.create", Map.of(
            "endpoint", "POST /api/nexus/matches/{id}/memory-shares",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Share selected ask_each_time memories as immutable snapshots with the other participant of an accepted match.",
            "proposedFields", List.of("memoryIds", "userConfirmed"),
            "sideEffects", Map.of(
                    "createsPublicContent", false,
                    "visibility", "accepted_match_participants_only",
                    "snapshot", true)
        ));

        c.put("memory_share.revoke", Map.of(
            "endpoint", "PATCH /api/nexus/matches/{matchId}/memory-shares/{shareId}",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Revoke one of the current user's memory snapshots from a match after explicit confirmation.",
            "proposedFields", List.of("revoked", "userConfirmed"),
            "sideEffects", Map.of(
                    "createsPublicContent", false,
                    "visibility", "accepted_match_participants_only")
        ));

        c.put("help_request.create", Map.of(
            "endpoint", "POST /api/nexus/help-requests",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Create a public help request after the user explicitly confirms the exact title, summary, labels, visibility, and safety notes.",
            "proposedFields", List.of("title", "summary", "content", "neededLabels", "categoryLabel",
                    "urgency", "locationHint", "meetingSafetyState", "agentContext", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", true,
                "createsPublicDiscussionAndHelpRequest", true,
                "visibility", "public_help_board",
                "serverAppendedFooter", HELP_FOOTER
            )
        ));

        c.put("dispatch.create", Map.of(
            "endpoint", "POST /api/nexus/help-requests/{id}/dispatches",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Dispatch a help request to a specific helper after the requester confirms.",
            "proposedFields", List.of("helperUserId", "message", "rationale", "meetingHint",
                    "meetingSafetyState", "expiresAt", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "sendsNotification", true,
                "visibility", "requester_and_helper_only"
            )
        ));

        c.put("match.create", Map.of(
            "endpoint", "POST /api/nexus/help-requests/{id}/matches",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Offer help on a help request after the helper explicitly confirms.",
            "proposedFields", List.of("message", "meetingHint", "meetingSafetyState", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", true,
                "postsInDiscussion", true,
                "visibility", "requester_and_helper_with_public_post",
                "serverAppendedFooter", MATCH_OFFER_FOOTER
            )
        ));

        c.put("match_message.create", Map.of(
            "endpoint", "POST /api/nexus/matches/{id}/messages",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Send a private coordination message within an accepted match after explicit confirmation. Use kind (chat|update|question|handoff) and clientRequestId for safe retries.",
            "proposedFields", List.of("content", "kind", "clientRequestId", "agentContext", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "idempotency", "clientRequestId replays return the original message"
            )
        ));

        c.put("match_task.create", Map.of(
            "endpoint", "POST /api/nexus/matches/{id}/tasks",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Add a shared task to an accepted match's collaboration plan so both sides can track progress.",
            "proposedFields", List.of("title", "note", "ownerRole", "clientRequestId", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "idempotency", "clientRequestId replays return the original task",
                "blockedWhilePaused", true
            )
        ));

        c.put("match_task.update", Map.of(
            "endpoint", "PATCH /api/nexus/matches/{id}/tasks/{taskId}",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Advance, block (with blockedReason), retitle, or hand over a shared collaboration task. Status changes are allowed only for the side that owns the task.",
            "proposedFields", List.of("status", "title", "note", "ownerRole", "blockedReason", "orderIndex", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "blockedWhilePaused", true
            )
        ));

        c.put("match_decision.create", Map.of(
            "endpoint", "POST /api/nexus/matches/{id}/decisions",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Open a human decision gate with 2-5 concrete options. assignedRole must be the counterpart (cross-party checkpoint); work that depends on the choice must wait until the assigned human decides.",
            "proposedFields", List.of("title", "context", "options", "assignedRole", "clientRequestId", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "idempotency", "clientRequestId replays return the original decision",
                "blockedWhilePaused", true
            )
        ));

        c.put("match_decision.resolve", Map.of(
            "endpoint", "PATCH /api/nexus/matches/{id}/decisions/{decisionId}",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Record the assigned human participant's choice (action=decide with optionKey) or, as the raiser, withdraw an open decision (action=cancel). Human gate: requires userConfirmed only, not allowAgentMatching.",
            "proposedFields", List.of("action", "optionKey", "note", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "idempotency", "re-deciding with the same optionKey returns the decided record"
            )
        ));

        c.put("match_deliverable.create", Map.of(
            "endpoint", "POST /api/nexus/matches/{id}/deliverables",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Submit the agreed deliverable (title, access hint, checksum, license note) for the counterpart's review.",
            "proposedFields", List.of("title", "description", "accessHint", "checksum", "licenseNote", "clientRequestId", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only",
                "idempotency", "clientRequestId replays return the original deliverable",
                "blockedWhilePaused", true
            )
        ));

        c.put("match_deliverable.review", Map.of(
            "endpoint", "PATCH /api/nexus/matches/{id}/deliverables/{deliverableId}",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Accept or reject (with reviewNote) a counterpart deliverable, or withdraw one's own pending deliverable. A pending deliverable blocks match completion. Human gate: requires userConfirmed only, not allowAgentMatching.",
            "proposedFields", List.of("action", "reviewNote", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only"
            )
        ));

        c.put("match_workspace.update", Map.of(
            "endpoint", "PATCH /api/nexus/matches/{id}/workspace",
            "requiresConfirmation", true,
            "permissions", List.of("allowAgentMatching"),
            "purpose", "Pause or resume collaboration, or move the baton (requester|helper|none). While the baton is set, only the holder can create tasks, decisions, and deliverables. Paused workspaces reject new work. Human control: requires userConfirmed only, not allowAgentMatching.",
            "proposedFields", List.of("collaborationState", "baton", "note", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "match_participants_only"
            )
        ));

        c.put("device_signal.create", Map.of(
            "endpoint", "POST /api/nexus/device-signals",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Send device signals (geohash, sensors) after explicit user confirmation.",
            "proposedFields", List.of("purpose", "coarseGeohash", "accuracyM", "bluetoothSeen",
                    "shakeDetected", "gyroAvailable", "payload", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "private_to_user"
            )
        ));

        c.put("llm_settings.update", Map.of(
            "endpoint", "PATCH /api/nexus/llm-settings",
            "requiresConfirmation", true,
            "permissions", List.of(),
            "purpose", "Update the current user's optional forum-side LLM provider metadata after confirmation.",
            "proposedFields", List.of("provider", "baseUrl", "apiKey", "chatModel", "responsesModel",
                    "supportsChatCompletions", "supportsResponses", "userConfirmed"),
            "sideEffects", Map.of(
                "createsPublicContent", false,
                "visibility", "private_to_user"
            )
        ));

        return c;
    }
}
